llvm*: Fix core usage & remove progress bar in tests

OPNA2608 opened the pull request:
###### Description of changes

[lit (LLVM Integrated Tester)](https://www.llvm.org/docs/CommandGuide/lit.html) by default uses as many threads as the build host
has cores, ignoring the user's core settings for the build.
This passes the configured core count on to lit, along with LLVM's default
settings for it [which we otherwise override in the process](https://github.com/llvm/llvm-project/blob/329fda39c507e8740978d10458451dcdb21563be/llvm/CMakeLists.txt#L559-L565).

Build command:
![Bildschirmfoto von 2022-08-27 20-00-57](https://user-images.githubusercontent.com/23431373/187047969-18f5b7d8-b571-48c1-9cf7-9804929670b6.png)

Before:
![Bildschirmfoto von 2022-08-27 19-03-30](https://user-images.githubusercontent.com/23431373/187047981-c56fb528-1ddc-42cf-adff-370fdb1016b3.png)

After:
![Bildschirmfoto von 2022-08-27 23-20-26](https://user-images.githubusercontent.com/23431373/187048423-7b73c360-e61a-43a5-98c7-2adc3adb3e15.png)

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  I test-built LLVM 5, 14 and 11 (default), they all use the correct amount of cores during `checkPhase` now.
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

primeos reviewed (changes requested):
Please also update `pkgs/development/compilers/llvm/git/` (important as this will be used as basis/template when packaging new LLVM major versions).
I think we should probably also set `--no-progress-bar` by default, if you don't mind. I suppose this doesn't work anyway since the Nix build doesn't run interactively and it might be required on Darwin?

Apart from that this LGTM, thanks :)
And thanks for the commit message (feel free to also include the URLs - I use Markdown references like in 20b2f59523a5afaf8f8bc011e945adc754874739 or d932886d6ea5b4e5bf07247e54c29f77b9b69a20 - but it's fine either way).

OPNA2608 committed:
llvm*: Fix core usage in tests

lit (LLVM Integrated Tester) [0] by default uses as many threads as the build host
has cores, ignoring the user's core settings for the build.
This passes the configured core count on to lit, along with LLVM's default
settings for it which we otherwise override in the process [1].

[0]: https://www.llvm.org/docs/CommandGuide/lit.html
[1]: https://github.com/llvm/llvm-project/blob/329fda39c507e8740978d10458451dcdb21563be/llvm/CMakeLists.txt#L559-L565

OPNA2608 commented:
Oh I missed the `git` one, thanks. :sweat_smile:

> I think we should probably also set `--no-progress-bar` by default, if you don't mind. I suppose this doesn't work anyway since the Nix build doesn't run interactively and it might be required on Darwin?

Can do. Test output after adding that looks ~ like this:

```
/nix/store/c1vb2z3c64i0sd92iz7fv0lb720qcvhb-python3-3.10.6/bin/python /build/llvm/utils/lit/lit.py -svj24 --no-progress-bar --param llvm_site_config=/build/llvm/build/test/lit.site.cfg --param llvm_unit_site_config=/build/llvm/build/test/Unit/lit.site.cfg /build/llvm/build>
-- Testing: 21448 tests, 24 threads --

Testing Time: 50.44s
  Expected Passes    : 20627
  Expected Failures  : 131
  Unsupported Tests  : 690
```

As opposed to the current:

```
/nix/store/c1vb2z3c64i0sd92iz7fv0lb720qcvhb-python3-3.10.6/bin/python /build/llvm/utils/lit/lit.py -svj24 --param llvm_site_config=/build/llvm/build/test/lit.site.cfg --param llvm_unit_site_config=/build/llvm/build/test/Unit/lit.site.cfg /build/llvm/build/test
                     -- Testing: 21448 tests, 24 threads --                     

^M  0% [-------------------------------------------------------------------------]
                                                                                ^M  0% [-------------------------------------------------------------------------]
LLVM-Unit :: ADT/./ADTTests/APFloatTest.FMA                                     ^M  0% [-------------------------------------------------------------------------]
LLVM-Unit :: ADT/./ADTTests/APFloatTest.DecimalStringsWithoutNullTerminators    ^M  0% [-------------------------------------------------------------------------]
LLVM-Unit :: ADT/./ADTTests/APFloatTest.MaxNum                                  ^M  0% [-------------------------------------------------------------------------]
LLVM-Unit :: ADT/./ADTTests/APFloatTest.Denormal                                ^M  0% [-------------------------------------------------------------------------]
LLVM-Unit :: ADT/./ADTTests/APFloatTest.MinNum                                  ^M  0% [-------------------------------------------------------------------------]

[...]

LLVM :: tools/llvm-objdump/macho-bad-bind.test                                  ^M 99% [==========================================================-] ETA: 00:00:00
LLVM :: tools/llvm-xray/X86/graph-diff-simple.txt                               ^M 99% [==========================================================-] ETA: 00:00:00
LLVM :: tools/sancov/covered_functions.test                                     ^M 99% [==========================================================-] ETA: 00:00:00
LLVM :: tools/sancov/print_coverage_pcs.test                                    ^M100% [===========================================================] ETA: 00:00:00
LLVM :: tools/sancov/blacklist.test                                             ^MTesting Time: 49.90s
  Expected Passes    : 20627
  Expected Failures  : 131
  Unsupported Tests  : 690
```

OPNA2608 committed:
llvm*: Don't show progress bar on tests

Since the progress bar results in large output that is broken due to the use of
ncurses and we already use the flag that gives verbose output on test failures,
let's just disable the progress bar.

primeos commented on pkgs/development/compilers/llvm/git/llvm/default.nix:
```diff
@@ -99,6 +99,13 @@ in stdenv.mkDerivation (rec {
+  preConfigure = ''
+    # Workaround for configure flags that need to have spaces
+    cmakeFlagsArray+=(
+      -DLLVM_LIT_ARGS='-svj''${NIX_BUILD_CORES} --no-progress-bar'
```
Strange, this is a bit surprising to me but IIRC the Nixpkgs abstractions do indeed often strip/replace quotes (among other processing steps or there's a quoting related bug in the code - Bash/shell quoting is unfortunately easy to get wrong...).

It'd be nice if we could avoid that preConfigure workaround but I guess not so that looks fine to me.

OPNA2608 replied:
The Nixpkgs manual [remarks about this detail](https://nixos.org/manual/nixpkgs/stable/#var-stdenv-configureFlagsArray) & [advises to use](https://nixos.org/manual/nixpkgs/stable/#var-stdenv-makeFlagsArray) `*FlagsArray` for these particular cases, `preConfigure` seemed like the best place for this.

primeos reviewed (approved):
Regarding `--no-progress-bar`: My bad, this does alter the output behaviour on Linux then but IMO it's still fine and likely even still desirable (the disadvantage is that watching the build output doesn't indicate the progress anymore but that information is only relevant when watching builds and AFAIK such output is often disabled in other places in Nixpkgs).

Anyway, LGTM now, thanks :)

primeos commented:
@Ericson2314 @sternenseemann does this look good to you as well (if you have the time for a brief look).

@ofborg build llvmPackages_git.stdenv

sternenseemann reviewed (approved):
Not sure if I'll miss the progress output though…
